### PR TITLE
Refactor co_ccap_parent_fee code

### LINKED ISSUE #3619: Refactor co_ccap_parent_fee code
https://github.com/PolicyEngine/policyengine-us/blob/master/policyengine_us/variables/gov/states/co/ccap/co_ccap_parent_fee.py

A few opportunities to simplify:

### Use scale parameters for the first and second multiplier

This is a marginal rate structure with respect to the FPG ratio. To get the absolute value, multiply by the FPG at the end.

https://github.com/PolicyEngine/policyengine-us/blob/8b48da58e3bbdc1ddc990090de17f23e8ba7a108/policyengine_us/variables/gov/states/co/ccap/co_ccap_parent_fee.py#L44-L52

### Use a scale parameter for discount rate

Start by getting to greatest quality rating in the SPM unit, then looking that up to get the discount factor. Can unify two parameters.

https://github.com/PolicyEngine/policyengine-us/blob/8b48da58e3bbdc1ddc990090de17f23e8ba7a108/policyengine_us/variables/gov/states/co/ccap/co_ccap_parent_fee.py#L85-L95

### PR DESCRIPTION
**Pull request by vrathi101**
Fixes #3619

### CODE REVIEW

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: run make format

File: policyengine_us/data/datasets/cps/cps.py [hunk 1]
```diff
@@ -416,17 +416,17 @@ def add_previous_year_income(self, cps: h5py.File) -> None:
     )
     df["in_sample"] = cps_cur_record_in_sample
     df["employment_income_prev"] = np.ones(len(df)) * np.nan
-    df["employment_income_prev"][cps_cur_record_in_sample] = (
-        cps_previous_year.loc[
-            cps_current_year.index[cps_cur_record_in_sample]
-        ].WSAL_VAL.values
-    )
+    df["employment_income_prev"][
```
Comment (PavelMakarchuk) [inline policyengine_us/data/datasets/cps/cps.py:419]: revert changes back

File: policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml [hunk 2]
```diff
@@ -1,12 +1,21 @@
-description: Colorado Child Care Assistance Program multiplies adjusted gross income or federal poverty guidelines with this factor to calculate base parent fee.
+description: Colorado Child Care Assistance Program multiplies adjusted gross income below or above federal poverty guidelines with these factors to calculate base parent fee.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml:1]:
```suggestion
description: Colorado Child Care Assistance Program multiplies adjusted gross income by the following factor to calculate base parent fee, based on gross income relative to the federal poverty guidelines.
```

File: policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml [hunk 3]
```diff
@@ -1,12 +1,21 @@
...
 metadata:
+  type: marginal_rate
+  threshold_unit: currency-USD
+  rate_unit: /1
+  label: Colorado Child Care Assistance Program multiplication factors
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml:6]:
```suggestion
  label: Colorado Child Care Assistance Program multiplication factor
```

File: policyengine_us/parameters/gov/states/co/ccap/is_quality_rating_discounted.yaml [hunk 4]
```diff
@@ -2,7 +2,7 @@ description: Whether Colorado Child Care Assistance Program povides discount whe
 metadata:
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/ccap/is_quality_rating_discounted.yaml:2]: change the description

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_parent_fee.py [hunk 5]
```diff
@@ -29,27 +29,21 @@ def formula(spm_unit, period, parameters):
...
-                gross_income * first_multiplication_factor,
-                fpg * first_multiplication_factor
-                + (gross_income - fpg) * second_multiplication_factor,
-            ),
-            2,
+        # Calculate base parent fee scaled (note income is monthly):
+        # When income_scaled <= 1: income_scaled * 0.01
+        # When income_scaled > 1: [1 * 0.01 + (income_scaled - 1) * 0.14]
+        # Multiply by fpg afterward to scale back up
+        gross_income_scaled = gross_income / fpg
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_parent_fee.py:42]:
```suggestion
        gross_income_fpg_ratio = gross_income / fpg
```

File: policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml:1]: How about reorganizing this to be `parent_fee/{base,add_on}.yaml`? where this is base

File: policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml [hunk 6]
```diff
@@ -1,12 +1,21 @@
-description: Colorado Child Care Assistance Program multiplies adjusted gross income or federal poverty guidelines with this factor to calculate base parent fee.
+description: Colorado Child Care Assistance Program multiplies adjusted gross income by the following factor to calculate base parent fee, based on gross income relative to the federal poverty guidelines.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml:1]: Please also change other variables (Python and PolicyEngine) and descriptions/labels to refer consistently to countable income or adjusted gross income (appears to be the former).
```suggestion
description: Colorado calculates its Child Care Assistance Program base parent fee by applying this rate schedule to the ratio of countable income to the federal poverty guideline, then multiplying the result by the federal poverty guideline.
```

File: policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml [hunk 7]
```diff
@@ -1,12 +1,21 @@
...
+description: Colorado Child Care Assistance Program multiplies adjusted gross income by the following factor to calculate base parent fee, based on gross income relative to the federal poverty guidelines.  
 metadata:
+  type: marginal_rate
+  threshold_unit: currency-USD
+  rate_unit: /1
+  label: Colorado Child Care Assistance Program multiplication factor
   reference:
     # This factor is included in the calculation of base parent fee in the spreadsheets
     - title: At 100% FPG Parent Fee Tool
       href: https://docs.google.com/spreadsheets/d/1EEc3z8Iwu_KRTlBtd2NssDDEx_FITqVq/edit#gid=468321263
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/co/ccap/base_parent_fee/first_and_second_multiplication_factor.yaml:10]: While you're here could you change the reference to publicly available documents?

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/parameters/gov/states/co/ccap/parent_fee/base.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+description: Colorado calculates its Child Care Assistance Program base parent fee by applying this rate schedule to the ratio of countable income to the federal poverty guideline, then multiplying the result by the federal poverty guideline.  
+metadata:
+  type: marginal_rate
+  threshold_unit: currency-USD
+  rate_unit: /1
+  label: Colorado Child Care Assistance Program multiplication factor
+  reference:
+    # This factor is included in the calculation of base parent fee
+    - title: COLORADO CHILD CARE ASSISTANCE PROGRAM RULES AND REGULATIONS - 3.111 PARENT FEES - P
```
[the remark this answers is not in the document]
Comment (vrathi101) [reply policyengine_us/parameters/gov/states/co/ccap/parent_fee/base.yaml]: Sure, should I remove references to spreadsheets in the CCCAP variables as well?

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_base_parent_fee.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        gross_income = spm_unit("co_ccap_countable_income", period)
+        # snap_fpg is monthly.
+        fpg = spm_unit("snap_fpg", period)
+        # Calculate base parent fee scaled (note income is monthly):
+        # When income_scaled <= 1: income_scaled * 0.01
+        # When income_scaled > 1: [1 * 0.01 + (income_scaled - 1) * 0.14]
+        # Multiply by fpg afterward to scale back up
+        gross_income_fpg_ratio = gross_income / fpg
+        base_parent_fee_scaled = p.parent_fee.base.calc(gross_income_fpg_ratio)
+        base_parent_fee = base_parent_fee_scaled * fpg
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_base_parent_fee.py:32]:
```suggestion
        return base_parent_fee_scaled * fpg
```

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        eligible_children = spm_unit("co_ccap_eligible_children", period)
+        # Calculate add-on parent fee based on the number of eligible
+        # children in a household and income:
+        # When income <= fpg: 0
+        # When income > fpg: 15 for each additional child
+        add_on_parent_fee_if_applies = (
+            eligible_children - 1
+        ) * p.parent_fee.add_on
+        add_on_parent_fee_applies = gross_income > fpg
+        add_on_parent_fee = (
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py:35]:
```suggestion
        return (
```

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        # Calculate base parent fee and add on parent fee.
+        gross_income = spm_unit("co_ccap_countable_income", period)
+        # snap_fpg is monthly.
+        fpg = spm_unit("snap_fpg", period)
+        eligible_children = spm_unit("co_ccap_eligible_children", period)
+        # Calculate add-on parent fee based on the number of eligible
+        # children in a household and income:
+        # When income <= fpg: 0
+        # When income > fpg: 15 for each additional child
+        add_on_parent_fee_if_applies = (
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py:31]:
```suggestion
        add_on_parent_fee_amount = (
```

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        year = period.start.year
+        if period.start.month >= 10:
+            instant_str = f"{year}-10-01"
+        else:
+            instant_str = f"{year - 1}-10-01"
+        p = parameters(instant_str).gov.states.co.ccap
+        # Calculate base parent fee and add on parent fee.
+        gross_income = spm_unit("co_ccap_countable_income", period)
+        # snap_fpg is monthly.
+        fpg = spm_unit("snap_fpg", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py:25]:
```suggestion
        snap_fpg = spm_unit("snap_fpg", period)
```

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_base_parent_fee.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        year = period.start.year
+        if period.start.month >= 10:
+            instant_str = f"{year}-10-01"
+        else:
+            instant_str = f"{year - 1}-10-01"
+        p = parameters(instant_str).gov.states.co.ccap
+        # Calculate base parent fee and add on parent fee.
+        gross_income = spm_unit("co_ccap_countable_income", period)
+        # snap_fpg is monthly.
+        fpg = spm_unit("snap_fpg", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_base_parent_fee.py:25]:
```suggestion
        snap_fpg = spm_unit("snap_fpg", period)
```

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_base_parent_fee.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        # Calculate base parent fee scaled (note income is monthly):
+        # When income_scaled <= 1: income_scaled * 0.01
+        # When income_scaled > 1: [1 * 0.01 + (income_scaled - 1) * 0.14]
+        # Multiply by fpg afterward to scale back up
+        gross_income_fpg_ratio = gross_income / fpg
+        base_parent_fee_scaled = p.parent_fee.base.calc(gross_income_fpg_ratio)
+        base_parent_fee = base_parent_fee_scaled * fpg
+        return base_parent_fee
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_base_parent_fee.py:33]:
```suggestion
```

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        # When income <= fpg: 0
+        # When income > fpg: 15 for each additional child
+        add_on_parent_fee_if_applies = (
+            eligible_children - 1
+        ) * p.parent_fee.add_on
+        add_on_parent_fee_applies = gross_income > fpg
+        add_on_parent_fee = (
+            add_on_parent_fee_if_applies * add_on_parent_fee_applies
+        )
+        return add_on_parent_fee
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py:38]:
```suggestion
```

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    reference = (
+        "https://www.sos.state.co.us/CCR/GenerateRulePdf.do?ruleVersionId=11042&fileName=8%20CCR%201403-1#page=41",
+        "https://www.sos.state.co.us/CCR/GenerateRulePdf.do?ruleVersionId=11042&fileName=8%20CCR%201403-1#page=62",
+    )
+    unit = USD
+    definition_period = MONTH
+
+    def formula(spm_unit, period, parameters):
+        year = period.start.year
+        if period.start.month >= 10:
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/ccap/co_ccap_add_on_parent_fee.py:17]: need to cover the different years with tests

File: policyengine_us/parameters/gov/states/co/ccap/parent_fee/base.yaml [hunk 17]
```diff
@@ -0,0 +1,19 @@
+description: Colorado calculates its Child Care Assistance Program base parent fee by applying this rate schedule to the ratio of countable income to the federal poverty guideline, then multiplying the result by the federal poverty guideline.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/ccap/parent_fee/base.yaml:1]:
```suggestion
description: Colorado multiplies the ratio of countable income to the federal poverty guideline by this rate under the Child Care Assistance Program to calculate the base parent fee.  
```

File: policyengine_us/parameters/gov/states/co/ccap/parent_fee/add_on.yaml [hunk 18]
```diff
@@ -0,0 +1,10 @@
+description: Colorado Child Care Assistance Program multiplies additional number of eligible children with this factor to calculate add on parent fee.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/ccap/parent_fee/add_on.yaml:1]:
```suggestion
description: Colorado multiplies the number of additional eligible children by this amount under the Child Care Assistance Program to calculate add on parent fee.
```

**Review by PavelMakarchuk — APPROVED** [review]

**Comment by PavelMakarchuk** [comment]
@MaxGhenis let us know if is more representative than the original code

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**